### ISSUE TITLE
Not using the specified user when using 

### ISSUE THREAD

**Issue by ifraixedes** [umputun/spot#129]
I have a task that has a multiline script command and it seems that `spot` doesn't use the provided username when running it.
I can observe that if the command is single-line it works fine, but the multiline not.

I think it isn't user the provided user because when I was using an SSH key that matched my local user, it worked fine, however I may be wrong.

This is a minimal `spot` playbook that fails, I changed the real host and username by some example ones for posting the issue.

```toml
ssh_key = ".tmp/ssh-key.priv"

[[targets.prod.hosts]]
host = "example.com"
user = "my-user"
name = "vm1"


## Task: Disable Jenkins node for maintenance
[[tasks]]
name = "task1"
targets = [ "prod" ]

[[tasks.commands]]
name = "single-line"
script = "pwd"

[[tasks.commands]]
name = "multi-line"
script = """
#!/bin/bash
pwd"""
```

Below the `spot` command that I ran and the output

```
spot -v --dbg -p test.toml -t prod --task task1
spot v1.8.1-7040925-2023-06-12T21:01:17Z
2023/07/14 20:33:13.944 [DEBUG] request to load playbook "test.toml"
2023/07/14 20:33:13.945 [INFO]  playbook loaded with 1 tasks
2023/07/14 20:33:13.945 [DEBUG] load command "single-line" (task: task1)
2023/07/14 20:33:13.945 [DEBUG] load command "multi-line" (task: task1)
2023/07/14 20:33:13.945 [INFO]  no inventory loaded
2023/07/14 20:33:13.945 [INFO]  ssh key: .tmp/ssh-key.priv
2023/07/14 20:33:13.945 [DEBUG] use private key ".tmp/ssh-key.priv"
2023/07/14 20:33:13.945 [DEBUG] task "task1" has 2 commands
2023/07/14 20:33:13.945 [DEBUG] target "prod" found in playbook
2023/07/14 20:33:13.945 [DEBUG] target "prod" has 1 hosts: [{Name:vm1 Host:example.com Port:0 User:my-user Tags:[]}]
2023/07/14 20:33:13.945 [DEBUG] target "prod" has 1 total hosts: [{Name:vm1 Host:example.com Port:0 User:my-user Tags:[]}]
2023/07/14 20:33:13.945 [DEBUG] target hosts (1) [{Name:vm1 Host:example.com Port:22 User:my-user Tags:[]}]
2023/07/14 20:33:13.945 [DEBUG] connect to "example.com:22" (vm1), user "my-user"
2023/07/14 20:33:13.945 [DEBUG] create ssh session to example.com:22, user my-user
2023/07/14 20:33:14.922 [DEBUG] ssh session created to example.com:22
[vm1 example.com:22] run task "task1", commands: 2
2023/07/14 20:33:14.923 [INFO]  run command "single-line" on host "example.com:22" (vm1)
2023/07/14 20:33:14.923 [DEBUG] execute script "single-line" on example.com:22
2023/07/14 20:33:14.923 [DEBUG] command "single-line" is single line, using script string
2023/07/14 20:33:14.923 [DEBUG] run sh -c 'pwd'
2023/07/14 20:33:14.923 [DEBUG] run ssh command "sh -c 'pwd'" on 35.222.11.8:22
[vm1 example.com:22]  > sh -c 'pwd'
[vm1 example.com:22]  > /home/my-user
[vm1 example.com:22] completed command "single-line" {script: sh -c 'pwd'} (380ms)
2023/07/14 20:33:15.304 [INFO]  run command "multi-line" on host "example.com:22" (vm1)
2023/07/14 20:33:15.304 [DEBUG] execute script "multi-line" on example.com:22
2023/07/14 20:33:15.304 [DEBUG] command "multi-line" is multiline, using script file
2023/07/14 20:33:15.304 [DEBUG] upload /tmp/spot-script2260163402 to /tmp/.spot/spot-script2260163402
2023/07/14 20:33:15.304 [DEBUG] upload /tmp/spot-script2260163402 to example.com:/tmp/.spot/spot-script2260163402
2023/07/14 20:33:15.639 [DEBUG] file mode for /tmp/spot-script2260163402: 0700
2023/07/14 20:33:16.086 [INFO]  uploaded /tmp/spot-script2260163402 to example.com:/tmp/.spot/spot-script2260163402 in 781.708961ms
[vm1 example.com:22]  ! failed command "multi-line" on host example.com:22 (vm1): can't prepare script on example.com:22: can't upload script to example.com:22: failed to create remote file: permission denied
2023/07/14 20:33:16.087 [ERROR] can't run task "task1" for target "prod": 1 error(s) occurred: [0] {failed command "multi-line" on host example.com:22 (vm1): can't prepare script on example.com:22: can't upload script to example.com:22: failed to create remote file: permission denied}
>>> stack trace:
main.main()
        /home/runner/work/spot/spot/cmd/spot/main.go:111 +0x272
panic: [ERROR] can't run task "task1" for target "prod": 1 error(s) occurred: [0] {failed command "multi-line" on host example.com:22 (vm1): can't prepare script on example.com:22: can't upload script to example.com:22: failed to create remote file: permission denied}

goroutine 1 [running]:
log.Panicf({0xd4f9c3?, 0x0?}, {0xc000223f60?, 0x9?, 0x7ffd269650ed?})
        /opt/hostedtoolcache/go/1.20.4/x64/src/log/log.go:391 +0x67
main.main()
        /home/runner/work/spot/spot/cmd/spot/main.go:111 +0x272
```

Thank you!

**Comment by umputun** [comment]
I can't reproduce it. This is my playbook:

```yml
user: umputun_bad
targets:
  dev:
    hosts:
      - {name: "dev1", host: "dev1.umputun.dev", port: 22, user: umputun}
      - {name: "dev2", host: "dev2.umputun.dev", port: 22, user: umputun_bad_too}

tasks:
  - name: run-things
    commands:
    - name: test
      script: |
        cd /srv
        dte=$(date +%Y%m%d)
        echo $dte
        echo "test" > /tmp/test-$dte.txt
```

```
 dev.umputun/spot [master|✔] spot -p spot-private.yml -v --dbg --target=dev
spot v1.9.0-480f8f5-2023-07-10T18:13:05Z
2023/07/14 16:37:22.260 [DEBUG] request to load playbook "spot-private.yml"
2023/07/14 16:37:22.261 [INFO]  playbook loaded with 1 tasks
2023/07/14 16:37:22.261 [DEBUG] load command "test" (task: run-things)
2023/07/14 16:37:22.261 [INFO]  no inventory loaded
2023/07/14 16:37:22.262 [INFO]  ssh key: /Users/umputun/.ssh/id_rsa
2023/07/14 16:37:22.262 [DEBUG] use private key "/Users/umputun/.ssh/id_rsa"
2023/07/14 16:37:22.262 [DEBUG] task "run-things" has 1 commands
2023/07/14 16:37:22.262 [DEBUG] target "dev" found in playbook
2023/07/14 16:37:22.262 [DEBUG] target "dev" has 2 hosts: [{Name:dev1 Host:dev1.umputun.dev Port:22 User:umputun Tags:[]} {Name:dev2 Host:dev2.umputun.dev Port:22 User:umputun_bad_too Tags:[]}]
2023/07/14 16:37:22.262 [DEBUG] target "dev" has 2 total hosts: [{Name:dev1 Host:dev1.umputun.dev Port:22 User:umputun Tags:[]} {Name:dev2 Host:dev2.umputun.dev Port:22 User:umputun_bad_too Tags:[]}]
2023/07/14 16:37:22.262 [DEBUG] target hosts (2) [{Name:dev1 Host:dev1.umputun.dev Port:22 User:umputun Tags:[]} {Name:dev2 Host:dev2.umputun.dev Port:22 User:umputun_bad_too Tags:[]}]
2023/07/14 16:37:22.262 [DEBUG] connect to "dev1.umputun.dev:22" (dev1), user "umputun"
2023/07/14 16:37:22.262 [DEBUG] create ssh session to dev1.umputun.dev:22, user umputun
2023/07/14 16:37:22.317 [DEBUG] ssh session created to dev1.umputun.dev:22
[dev1 dev1.umputun.dev:22] run task "run-things", commands: 1
2023/07/14 16:37:22.317 [INFO]  run command "test" on host "dev1.umputun.dev:22" (dev1)
2023/07/14 16:37:22.317 [DEBUG] execute script "test" on dev1.umputun.dev:22
2023/07/14 16:37:22.317 [DEBUG] command "test" is multiline, using script file
2023/07/14 16:37:22.317 [DEBUG] upload /var/folders/w5/dsr8pn3d4qqbyybsr_9vhc1r0000gn/T/spot-script368381278 to /tmp/.spot/spot-script368381278
2023/07/14 16:37:22.317 [DEBUG] upload /var/folders/w5/dsr8pn3d4qqbyybsr_9vhc1r0000gn/T/spot-script368381278 to dev1.umputun.dev:/tmp/.spot/spot-script368381278
2023/07/14 16:37:22.320 [DEBUG] file mode for /var/folders/w5/dsr8pn3d4qqbyybsr_9vhc1r0000gn/T/spot-script368381278: 0700
2023/07/14 16:37:22.323 [INFO]  uploaded /var/folders/w5/dsr8pn3d4qqbyybsr_9vhc1r0000gn/T/spot-script368381278 to dev1.umputun.dev:/tmp/.spot/spot-script368381278 in 5.726417ms
2023/07/14 16:37:22.323 [DEBUG] run sh -c /tmp/.spot/spot-script368381278
2023/07/14 16:37:22.323 [DEBUG] run ssh command "sh -c /tmp/.spot/spot-script368381278" on 192.168.1.41:22
[dev1 dev1.umputun.dev:22]  > sh -c /tmp/.spot/spot-script368381278
[dev1 dev1.umputun.dev:22]  > 20230714
2023/07/14 16:37:22.329 [INFO]  deleted /tmp/.spot/spot-script368381278
[dev1 dev1.umputun.dev:22] completed command "test" {script: sh -c /tmp/.spot/spot-script368381278} (12ms)
[dev1 dev1.umputun.dev:22] script: /tmp/.spot/spot-script368381278
[dev1 dev1.umputun.dev:22]  + #!/bin/sh
[dev1 dev1.umputun.dev:22]  + set -e
[dev1 dev1.umputun.dev:22]  + cd /srv
[dev1 dev1.umputun.dev:22]  + dte=$(date +%Y%m%d)
[dev1 dev1.umputun.dev:22]  + echo $dte
[dev1 dev1.umputun.dev:22]  + echo "test" > /tmp/test-$dte.txt
[dev1 dev1.umputun.dev:22] completed task "run-things", commands: 1 (67ms)
2023/07/14 16:37:22.330 [DEBUG] connect to "dev2.umputun.dev:22" (dev2), user "umputun_bad_too"
2023/07/14 16:37:22.330 [DEBUG] create ssh session to dev2.umputun.dev:22, user umputun_bad_too
[dev2 dev2.umputun.dev:22]  ! can't connect to dev2: failed to create client connection to dev2.umputun.dev:22: ssh: handshake failed: ssh: unable to authenticate, attempted methods [none publickey], no supported methods remain
2023/07/14 16:37:22.370 [ERROR] can't run task "run-things" for target "dev": 1 error(s) occurred: [0] {can't connect to dev2: failed to create client connection to dev2.umputun.dev:22: ssh: handshake failed: ssh: unable to authenticate, attempted methods [none publickey], no supported methods remain}
>>> stack trace:
main.main()
        /home/runner/work/spot/spot/cmd/spot/main.go:113 +0x21c
panic: [ERROR] can't run task "run-things" for target "dev": 1 error(s) occurred: [0] {can't connect to dev2: failed to create client connection to dev2.umputun.dev:22: ssh: handshake failed: ssh: unable to authenticate, attempted methods [none publickey], no supported methods remain}
```

As you can see, in my example, i have the wrong user defined globally and a correct user set for dev1. The attempt to connect to dev2 failed as expected, and you can see in this case, the error will indicate ssh issue clearly with ": ssh: handshake failed: ssh: unable to authenticate."

Do you have permission issues limiting writing to /tmp directory?

**Comment by ifraixedes** [comment]
I tried to change the host and the user in your `spot` file and the problem is the same.

The user shouldn't have any permission limit on the /tmp directory because if I establish an SSH connection with that user and key, I can write in `/tmp`.

I read the `/var/log/auth.log` in the server and I can confirm `spot` is connecting with the correct user.

This is very weird. I guess some permission limiting is going on, but I don't know why.

**Comment by umputun** [comment]
You can try to [change the temp directory](https://github.com/umputun/spot/blob/master/pkg/runner/commands.go#L35) and rebuild it to make sure this is something that is /tmp related.

**Comment by ifraixedes** [comment]
I found the issue. 
`spot` creates a `/tmp/.spot` directory with only write permission on the owner (which is the used user to connect o to the machine), it doesn't delete the directory when it finishes, and it doesn't recreate it in subsequent executions.

I connected to the machine and I deleted `/tmp/.spot` directory and then the other user can execute the task without issues.

**Comment by umputun** [comment]
Ah, it makes sense, indeed. So, the case you have is /tmp/.spot created for user A and after this, you run the command with user B.

Removing /tmp/.spot directory after the command execution is probably the right thing to do, but this can be unsafe if concurrent playbooks are running simultaneously. Alternatively, it can make this directory with 0x777 permission. What do you think?

**Comment by umputun** [comment]
and one more possible solution - making (and removing) a unique directory, i.e. /tmp/.spot-${rand}

**Comment by ifraixedes** [comment]
I had the same ideas.
I like the second, creating a unique directory, however, would be possible to create the directory emitting `mktemp -d .spot-xxxxx` rather than compute the random value by spot on the local?

I feel that it would be more unlikely to have clashes, however, I have to admit that using `crypto/rand` package should not cause any clash either.

**Comment by knutov** [comment]
Something like `mktemp -d --suffix=.spot` is definitely better, I think

**Comment by umputun** [comment]
Invoking remote command for temp directory creation is not easy, as we need to communicate this path back to the spot. It is also another platform-specific thing I'm trying to avoid. So, the change I did in #130 makes this random part of /tmp/.spot-{rand} and cleaning on completion. It is relevant not just for the multiline script but for copy with the sudo flag as well. Both cases should be fine now and deployed to 1.9.1.

Pls check it out and confirm if it fixed your problem.

**Comment by ifraixedes** [comment]
I can confirm that patch #130 released on [v1.9.1](https://github.com/umputun/spot/releases/tag/v1.9.1) works.
